tools/gen_customop_template.py

```diff
@@ -0,0 +1,109 @@
+'''
+Generates the C++ template code for a new custom operator.
+
+Usage: $ python tools\gen_customop_template.py onnx_model_filepath customop_template_filepath
+'''
+
+import argparse
+import onnx
+import pathlib
+
+# Load and check ONNX model
+def load_onnx_model(model_path):
+    model = onnx.load(model_path)
+    onnx.checker.check_model(model)
+    return model
+
+# Get input and output type count from the ONNX model using protobuf information
+def get_io_count(model):
+    input_count = 0
+    output_count = 0
+    custom_op_node_exists = False
+
+    print("Note: This C++ CustomOp template generator currently only supports models with one custom op node.\n________\n")
+    for node in model.graph.node:
+        # Find CustomOp node using domain
+        if node.domain == "ai.onnx.contrib" or node.domain == "com.microsoft.extensions":
+            custom_op_node_exists = True
+            input_count = len(node.input)
+            output_count = len(node.output)
+
+    if not custom_op_node_exists:
+        sys.exit("Error: No custom op node present in provided model")
+
+    return input_count, output_count
+
+# Add initial CustomOp code to C++ header file for CustomOp template
+def create_hpp(customop_template_filepath, op, op_name, input_type_count, output_type_count):
+    customop_template = r'''
+    #include <stdio.h>
+    #include "ocos.h"
+    #include "string_utils.h"
+    #include "ustring.h"
+
+    void* {custom_op}::CreateKernel(const OrtApi& api, const OrtKernelInfo* info) const {{
+      return CreateKernelImpl(api, info);
+    }};
+
+    const char* {custom_op}::GetName() const {{ return "{op_name}"; }};
+
+    size_t {custom_op}::GetInputTypeCount() const {{
+      return {input_type_count};
+    }};
+
+    // Note: the following method is not complete and contains a temporary default return value.
+    // Change return value to appropriate data type based on ONNXTensorElementDataType
+    // mapping of TensorProto.DataType value
+    ONNXTensorElementDataType {custom_op}::GetInputType(size_t /*index*/) const {{
+      return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+    }};
+
+    size_t {custom_op}::GetOutputTypeCount() const {{
+      return {output_type_count};
+    }};
+
+    // Note: the following method is not complete and contains a temporary default return value.
+    // Change return value to appropriate data type based on ONNXTensorElementDataType
+    // mapping of TensorProto.DataType value
+    ONNXTensorElementDataType {custom_op}::GetOutputType(size_t /*index*/) const {{
+      return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+    }};
+    '''
+
+    hpp_file = customop_template.format(custom_op = op, op_name = op_name, input_type_count = input_type_count, output_type_count = output_type_count)
+
+    # Write code to C++ template file
+    with open(customop_template_filepath,'wt') as file:
+        print("Added C++ CustomOp template code to output filepath: " + str(customop_template_filepath) + "\n")
```

Comment:
print() will add a newline at the end by default.
f-strings are a convenient shorthand (https://docs.python.org/3/reference/lexical_analysis.html#f-strings):
```suggestion
        print(f"Added C++ CustomOp template code to output filepath: {customop_template_filepath}")
```

Reply:
right, missed replacing that with a f-string, thanks!

Reply:
given ORT stops support of Python < 3.6, we can switch to f-string now